lib/model.js

```diff
@@ -176,62 +182,99 @@ Model.new = function(name, schema, options, thinky) {
   return model
 }
 
-Model.prototype.isReady = function() {
-  return this._getModel()._tableReady;
-}
+/**
+ * Create the model's table.
+ * @return {Promise=} Returns a promise which will resolve when the table is ready.
+ */
+Model.prototype.tableReady = function() {
+  var self = this;
+  var model = this._getModel();
+  if (!this._initModel) return Promise.resolve();
+  if (this._tableReadyPromise) return this._tableReadyPromise;
+
+  // Create the table, or push the table name in the queue.
+  var r = model._thinky.r;
+  this._tableReadyPromise = model._thinky.dbReady()
+  .then(function() {
+    return r.tableList().contains(model._name).do(function(result) {
+      return r.branch(
+        result,
+        {created: 0},
+        r.tableCreate(model._name, {primaryKey: model._pk})
+      );
+    }).run();
+  })
+  .error(function(error) {
+    model._error = error;
+    // Should we throw here?
+  });
 
-Model.prototype._setError = function(error) {
-  this._getModel()._error = error;
-  this.emit('error', error);
-}
+  return this._tableReadyPromise.then(function() {
+    self.emit('created');
+    if (!self._pendingPromises.length) {
+      self.emit('ready');
+    }
+  });
+};
 
-Model.prototype._setReady = function() {
-  this._getModel()._tableReady = true;
-  this.emit('ready', this);
-  while (this._getModel()._onTableReady.length > 0) {
-    this._getModel()._onTableReady.shift()();
-  }
-}
+/**
+ * Get a promise which resolves when the Model's table and
+ * all indices have been created.
+ */
+Model.prototype.ready = function() {
+  var requirements = [];
 
-Model.prototype._indexWasCreated = function(type) {
-  if (type === 'local') {
-    this._getModel()._indexesToCreate--;
-  }
-  else if (type === 'foreign') {
-    this._getModel()._foreignIndexesToCreate--;
-  }
-  // We need to check tableCreated because _indexWasCreated can be called when a foreign index is built.
-  if ((this._getModel()._tableCreated === true)
-    && (this._getModel()._indexesToCreate === 0)
-    && (this._getModel()._foreignIndexesToCreate === 0)
-    && (this._getModel()._indexesFetched === true)) {
+  // Ensure the Model's table is ready
+  requirements.push(this.tableReady());
 
-    this._setReady();
-  }
-}
+  // Ensure all other pending promises have been resolved
+  requirements.push(this._promisesReady());
 
+  return Promise.all(requirements);
+};
 
-Model.prototype._tableWasCreated = function() {
+Model.prototype._promisesReady = function() {
   var self = this;
+  if (this._promisesReadyPromise) return this._promisesReadyPromise;
+
+  var verifyAll = function() {
+    return Promise.all(self._pendingPromises)
+    .then(function() {
+      var i, allFullfilled = true;
+      for (i=0; i<self._pendingPromises.length; i++) {
+         if (!self._pendingPromises[i].isFulfilled()) {
+          allFullfilled = false;
+          break;
+         }
+      }
+      return allFullfilled ? Promise.resolve() : verifyAll();
```

Comment:
After taking a closer look, I'm confused as to what this is doing. The `for` loop seems redundant since `Promise.all` will only resolve once all the pending promises are fulfilled. @rasapetter did you mean to use `Promise.settle` instead or am I missing something?